### ISSUE TITLE
Multiple expressions

### ISSUE THREAD

**Issue by wonderful-panda** [PowerAssert/PowerAssert.Net#27]
It would be nice if two or more expressions can be tested at once.

For example,

```csharp
var test = "This is test.";
PAssert.AreTrue(
    () => test.Length == 10,
    () => test.StartsWith("This"),
    () => test.EndsWith("test")
);
```

```
System.Exception: AreTrue failed, 
--------------------------------------------------------------------------------
test.Length == 10
.  . .    . __
.  . \_ __/ |
\ _/   |    |
 |     |    False
 |     13
 "This is test."
--------------------------------------------------------------------------------
test.EndsWith("test")
.  . \__ ___/
\ _/    |
 |      False
 "This is test."
--------------------------------------------------------------------------------

   at ...
```

Advantage compared to below code is simple assertion message. This shows which part is actually wrong clearly.

```csharp
PAssert.IsTrue(() => test.Length == 10 && test.StartsWith("This") && test.EndsWith("test"));
```

Advantage compared to below code is coverage. All expressions are tested every time even when first one fails.

```csharp
PAssert.IsTrue(() => test.Length == 10);
PAssert.IsTrue(() => test.StartsWith("This"));
PAssert.IsTrue(() => test.EndsWith("test"));
```

**Comment by Porges** [comment]
This makes sense to me. Just to check, you're assuming that all forms force all arguments?

**Comment by wonderful-panda** [comment]
Sorry, I didn't understand your question.
Could you tell me more about ?

**Comment by robfe** [comment]
I've just read https://dannorth.net/2016/09/03/scratching-a-junit-itch/. I'm playing with something that would behave like that, but probably be a bit more powerful. It might end up scratching the itch that you have here

**Comment by robfe** [comment]
This is in https://github.com/PowerAssert/PowerAssert.Net/commit/f80332606d9489dca72e5bd72f29ec5380a249ef - what do you think?

**Comment by wonderful-panda** [comment]
Looks nice.

The only one concern is that stack traces of each failures have little or no information.
I prefer that they are omitted like below.

```
﻿PolyAssert encountered 3 failures:
 
    at PowerAssert.MultipleAssertions.PolyAssert.StopIfErrorsHaveOccurred() in ...\PolyAssert.cs
    at PowerAssert.MultipleAssertions.PolyAssert.Dispose() in ...\PolyAssert.cs
    at PowerAssertTests.Approvals.EndToEndTest.PolyAssert() in ...\EndToEndTest.cs
 
 ============================================================
 
 > I will report all the errors i saw when i get disposed
 
 ============================================================
IsTrue failed: 

 x == 5
 . __
 . |
 | False
 6
 ============================================================
 
 ============================================================
IsTrue failed: 

 x == 7
 . __
 . |
 | False
 6
 ============================================================
 
 ============================================================
 NUnit.Framework.AssertionException: Wah wah
 ============================================================
 
 > PolyAssert.Log messages are only printed if the test fails
```

**Comment by robfe** [comment]
That's a really good point @wonderful-panda. But would it be useful to show the first line of the stacktrace? That tells you which line in the test the failing assertion is on...

**Comment by wonderful-panda** [comment]
It would be useful if stacktrace shows user code location (e.g. `at ... in .../EndToEndTest.cs: line 484`).
But now, it only shows location in this library (`in .../PAssert.cs`).

That should be fixed separately from this issue.
It would be nice if output is like below
```
IsTrue failed (at ...PolyAssert() in .../EndToEndTest.cs: line 484) : 

 x == 5
 . __
 . |
 | False
 6
```

**Comment by robfe** [comment]
Great thinking - how's https://github.com/PowerAssert/PowerAssert.Net/commit/8744f33800e4f5905c327195732c10804a21890c

**Comment by wonderful-panda** [comment]
Unfortunately, I still have some v4.0 projects.
how about this instead of `[CallerFilePath]` and `[CallerLineNo]` ?

``` csharp
var asm = Assembly.GetExecutingAssembly();
var st = new StackTrace(1, true);
var callerFrame = st.GetFrames().FirstOrDefault(f => f.GetMethod().DeclaringType?.Assembly != asm);

var path = callerFrame?.GetFileName();
var lineno = callerFrame?.GetFileLineNumber();
```

(Of course, feel free to drop 4.0 support if you want)

**Comment by robfe** [comment]
Hmm. I already have that code for another purpose. I'll give that a go too tomorrow

**Comment by robfe** [comment]
@wonderful-panda it's done your way now. Thanks for the idea!

**Comment by wonderful-panda** [comment]
Thank you!

**Comment by wonderful-panda** [comment]
I found a small problem. Please check #29 out.

**Comment by robfe** [comment]
merged #29
